File: graph/BellmanFord/cpp/BellmanFord.cpp
#include "BellmanFord.hpp"

#include <limits>

namespace {

bool CheckedAdd(long long left, long long right, long long& out) {
    if (right > 0 && left > std::numeric_limits<long long>::max() - right) {
        return false;
    }

    if (right < 0 && left < std::numeric_limits<long long>::min() - right) {
        return false;
    }

    out = left + right;
    return true;
}

}  // namespace

BellmanFordResult BellmanFord(int nodeCount, const std::vector<Edge>& edges, int start) {
    BellmanFordResult result{BellmanFordStatus::Ok, std::vector<std::optional<long long>>(nodeCount)};
    result.distances[start] = 0;

    for (int pass = 0; pass < nodeCount - 1; ++pass) {
        bool updated = false;

        for (const Edge& edge : edges) {
            if (!result.distances[edge.from].has_value()) {
                continue;
            }

            long long candidate = 0;
            if (!CheckedAdd(*result.distances[edge.from], edge.weight, candidate)) {
                result.status = BellmanFordStatus::Overflow;
                return result;
            }

            if (!result.distances[edge.to].has_value() || candidate < *result.distances[edge.to]) {
                result.distances[edge.to] = candidate;
                updated = true;
            }
        }

        if (!updated) {
            break;
        }
    }

    for (const Edge& edge : edges) {
        if (!result.distances[edge.from].has_value()) {
            continue;
        }

        long long candidate = 0;
        if (!CheckedAdd(*result.distances[edge.from], edge.weight, candidate)) {
            result.status = BellmanFordStatus::Overflow;
            return result;
        }

        if (!result.distances[edge.to].has_value() || candidate < *result.distances[edge.to]) {
            result.status = BellmanFordStatus::NegativeCycle;
            return result;
        }
    }

    return result;
}

File: graph/BellmanFord/rust/src/bellman_ford.rs
#[derive(Clone, Copy)]
pub struct Edge {
    pub source: usize,
    pub target: usize,
    pub weight: i64,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Status {
    Ok,
    NegativeCycle,
    Overflow,
}

pub struct Result {
    pub status: Status,
    pub distances: Vec<Option<i64>>,
}

pub fn bellman_ford(node_count: usize, edges: &[Edge], start: usize) -> Result {
    let mut distances = vec![None; node_count];
    distances[start] = Some(0_i64);

    for _ in 0..node_count.saturating_sub(1) {
        let mut updated = false;

        for edge in edges {
            let Some(source_distance) = distances[edge.source] else {
                continue;
            };

            let Some(candidate) = source_distance.checked_add(edge.weight) else {
                return Result {
                    status: Status::Overflow,
                    distances,
                };
            };

            match distances[edge.target] {
                Some(target_distance) if candidate >= target_distance => {}
                _ => {
                    distances[edge.target] = Some(candidate);
                    updated = true;
                }
            }
        }

        if !updated {
            break;
        }
    }

    for edge in edges {
        let Some(source_distance) = distances[edge.source] else {
            continue;
        };

        let Some(candidate) = source_distance.checked_add(edge.weight) else {
            return Result {
                status: Status::Overflow,
                distances,
            };
        };

        match distances[edge.target] {
            Some(target_distance) if candidate >= target_distance => {}
            _ => {
                return Result {
                    status: Status::NegativeCycle,
                    distances,
                };
            }
        }
    }

    Result {
        status: Status::Ok,
        distances,
    }
}
